src/types/mutex.rs

```diff
@@ -0,0 +1,430 @@
+/// which is not `Send`. With `Rc`, we can have multiple copies pointing to the same heap
+/// allocation with a non-atomic reference count. If we were to use `Mutex<Rc<_>>`, it would
+/// only protect one instance of `Rc` from shared access, leaving other copies vulnerable
+/// to potential data races.
+///
+/// Also note that it is not necessary for `T` to be `Sync` as `&T` is only made available
+/// to one thread at a time if `T` is not `Sync`.
+///
+/// [`Rc`]: std::rc::Rc
+unsafe impl<T: ?Sized + Send> Sync for PyMutex<T> {}
+
+impl<T> PyMutex<T> {
+    /// Acquire the mutex, blocking the current thread until it is able to do so.
+    pub fn lock(&self) -> LockResult<PyMutexGuard<'_, T>> {
+        unsafe { crate::ffi::PyMutex_Lock(UnsafeCell::raw_get(&self.mutex)) };
+        PyMutexGuard::new(self)
+    }
+
+    /// Create a new mutex in an unlocked state ready for use.
+    pub const fn new(value: T) -> Self {
+        Self {
+            mutex: UnsafeCell::new(crate::ffi::PyMutex::new()),
+            data: UnsafeCell::new(value),
+            poison: Flag::new(),
+        }
+    }
+
+    /// Check if the mutex is locked.
+    ///
+    /// Note that this is only useful for debugging or test purposes and should
+    /// not be used to make concurrency control decisions, as the lock state may
+    /// change immediately after the check.
+    #[cfg(Py_3_14)]
+    pub fn is_locked(&self) -> bool {
+        let ret = unsafe { crate::ffi::PyMutex_IsLocked(UnsafeCell::raw_get(&self.mutex)) };
+        ret != 0
+    }
+
+    /// Consumes this mutex, returning the underlying data.
+    ///
+    /// # Errors
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then
+    /// this call will return an error containing the underlying data
+    /// instead.
+    pub fn into_inner(self) -> LockResult<T>
+    where
+        T: Sized,
+    {
+        let data = self.data.into_inner();
+        map_result(self.poison.borrow(), |()| data)
+    }
+
+    /// Clear the poisoned state from a mutex.
+    ///
+    /// If the mutex is poisoned, it will remain poisoned until this function is called. This
+    /// allows recovering from a poisoned state and marking that it has recovered. For example, if
+    /// the value is overwritten by a known-good value, then the mutex can be marked as
+    /// un-poisoned. Or possibly, the value could be inspected to determine if it is in a
+    /// consistent state, and if so the poison is removed.
+    pub fn clear_poison(&self) {
+        self.poison.clear();
+    }
+}
+
+#[cfg_attr(not(panic = "unwind"), allow(unnecessary-wraps))]
+fn map_result<T, U, F>(result: LockResult<T>, f: F) -> LockResult<U>
+where
+    F: FnOnce(T) -> U,
+{
+    match result {
+        Ok(t) => Ok(f(t)),
+        #[cfg(panic = "unwind")]
+        Err(e) => Err(PoisonError::new(f(e.into_inner()))),
+        #[cfg(not(panic = "unwind"))]
+        Err(_) => {
+            unreachable!();
```

Comment:
not sure if unreachable!() is right here. I don't think it's possible to hit this error if the target doesn't support unwinding though...